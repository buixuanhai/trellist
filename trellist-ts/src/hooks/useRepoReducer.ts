import produce from 'immer';
import { useReducer } from 'react';
import { Card, Repo } from '../components/types';

const reorder = (list: Card[], startIndex: number, endIndex: number) => {
  const [removed] = list.splice(startIndex, 1);
  list.splice(endIndex, 0, removed);
  return list;
};

type MoveCardBetweenListAction = {
  type: 'MOVE_CARD_BETWEEN_LIST';
  source: {
    index: number;
    droppableId?: string;
  };
  destination: {
    index: number;
    droppableId?: string;
  };
};

type AddCardAction = {
  type: 'ADD_CARD';
  listId: string;
  card: Card;
};

type DeleteCardAction = {
  type: 'DELETE_CARD';
  card: Card;
};

type UpdateCardAction = {
  type: 'UPDATE_CARD';
  card: Card;
};

type ReorderCardInList = {
  type: 'REORDER_CARD_IN_LIST';
  source: {
    index: number;
    droppableId: string;
  };
  destination: {
    index: number;
    droppableId: string;
  };
};
export type RepoAction =
  | MoveCardBetweenListAction
  | AddCardAction
  | DeleteCardAction
  | UpdateCardAction
  | ReorderCardInList;

export function reducer(state: Repo, action: RepoAction) {
  switch (action.type) {
    case 'MOVE_CARD_BETWEEN_LIST': {
      const {
        source: { index: sourceCardIndex, droppableId: sourceListId },
        destination: {
          index: destinationCardIndex,
          droppableId: destinationListId
        }
      } = action;

      return produce(state, draftState => {
        const sourceListIndex = draftState.lists.findIndex(
          l => l.id === sourceListId
        );
        const destinationListIndex = draftState.lists.findIndex(
          l => l.id === destinationListId
        );

        // take out card
        const card = draftState.lists[sourceListIndex].cards[sourceCardIndex];
        card.stage = draftState.lists[destinationListIndex].stage;

        // delete from old list
        draftState.lists[sourceListIndex].cards.splice(sourceCardIndex, 1);

        // insert into new list
        draftState.lists[destinationListIndex].cards.splice(
          destinationCardIndex,
          0,
          card
        );
        return draftState;
      });
    }
    case 'ADD_CARD': {
      const { listId, card } = action;
      return produce(state, draftState => {
        const listIndex = draftState.lists.findIndex(l => l.id === listId);
        draftState.lists[listIndex].cards.push({
          ...card,
          stage: draftState.lists[listIndex].stage
        });
        return draftState;
      });
    }
    case 'DELETE_CARD': {
      const { card } = action;
      return produce(state, draftState => {
        const listIdnex = draftState.lists.findIndex(
          l => l.stage === card.stage
        );
        const cardIndex = draftState.lists[listIdnex].cards.findIndex(
          c => c.id === card.id
        );
        draftState.lists[listIdnex].cards.splice(cardIndex, 1);
        return draftState;
      });
    }
    case 'UPDATE_CARD': {
      const { card } = action;
      return produce(state, draftState => {
        const listIndex = draftState.lists.findIndex(
          l => l.stage === card.stage
        );
        const cardIndex = draftState.lists[listIndex].cards.findIndex(
          c => c.id === card.id
        );
        draftState.lists[listIndex].cards[cardIndex] = { ...card };
        return draftState;
      });
    }
    case 'REORDER_CARD_IN_LIST': {
      const {
        source: { index: sourceIndex, droppableId: listId },
        destination: { index: destinationIndex }
      } = action;
      return produce(state, draftState => {
        const listIndex = draftState.lists.findIndex(l => l.id === listId);
        draftState.lists[listIndex].cards = reorder(
          draftState.lists[listIndex].cards,
          sourceIndex,
          destinationIndex
        );
      });
    }
    default:
      return state;
  }
}

export default function useRepoReducer(repo: Repo) {
  return useReducer(reducer, repo);
}
