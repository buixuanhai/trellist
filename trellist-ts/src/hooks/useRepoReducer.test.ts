import { Droppable } from "react-beautiful-dnd";
import useRepoReducer, { reducer, RepoAction } from "./useRepoReducer";


const repo = {
  "name": "Repo 1",
  "id": "1",
  "lists": [
    {
      "title": "Open",
      "stage": "open",
      "cards": [
        {
          "text": "Vulnerability 1",
          "id": "w",
          "stage": "open"
        },
        {
          "text": "Vulnerability 2",
          "id": "x",
          "stage": "open"
        }
      ],
      "id": "a"
    },
    {
      "title": "Confirmed",
      "stage": "confirmed",
      "cards": [
        {
          "text": "Vulnerability 3",
          "id": "y",
          "stage": "confirmed"
        },
        {
          "text": "Vulnerability 4",
          "id": "z",
          "stage": "confirmed"
        }
      ],
      "id": "b"
    },
    {
      "title": "False Positive",
      "stage": "falsePositive",
      "cards": [
        {
          "text": "Vulnerability 5",
          "id": "m",
          "stage": "falsePositive"
        },
        {
          "text": "Vulnerability 6",
          "id": "n",
          "stage": "falsePositive"
        }
      ],
      "id": "c"
    },
    {
      "title": "Fixed",
      "stage": "fixed",
      "cards": [
        {
          "text": "Vulnerability 7",
          "id": "o",
          "stage": "fixed"
        },
        {
          "text": "Vulnerability 8",
          "id": "p",
          "stage": "fixed"
        }
      ],
      "id": "d"
    }
  ]
}

describe('useRepoReducer', () => {
  test('test move card between lists', () => {
    const action: RepoAction = {
      type: 'MOVE_CARD_BETWEEN_LIST',
      source: {
        index: 0,
        droppableId: 'a'
      },
      destination: {
        index: 0,
        droppableId: 'd'
      }
    }

    const state = reducer(repo, action);
    expect(state).toMatchSnapshot()
  });

  test('move card in list', () => {
    const action: RepoAction = {
      type: 'REORDER_CARD_IN_LIST',
      source: {
        index: 0,
        droppableId: 'a'
      },
      destination: {
        index: 1,
        droppableId: 'a'
      }
    }

    const state = reducer(repo, action);
    expect(state.lists[0]).toMatchSnapshot()
  });
  test('add card', () => {
    const action: RepoAction = {
      type: 'ADD_CARD',
      listId: 'b',
      card: {
        id: 't',
        text: 'new card',
        stage: 'confirmed'
      }
    }
    const state = reducer(repo, action);
    expect(state.lists[1]).toMatchSnapshot()
  });

  test('delete card', () => {
    const action: RepoAction = {
      type: "DELETE_CARD",
      card: {
        "text": "Vulnerability 4",
        "id": "z",
        "stage": "confirmed"
      }
    }
    const state = reducer(repo, action);
    expect(state.lists[1]).toMatchSnapshot()
  });

  describe('update card', () => {
    test('update card title', () => {
      const action: RepoAction = {
        type: "UPDATE_CARD",
        card: {
          "text": "Vulnerability 4 (UPDATED)",
          "id": "z",
          "stage": "confirmed"
        }
      }
      const state = reducer(repo, action);
      expect(state.lists[1]).toMatchSnapshot()
    });

    test('update card note ', () => {
      const action: RepoAction = {
        type: "UPDATE_CARD",
        card: {
          "text": "Vulnerability 4",
          "id": "z",
          "stage": "confirmed",
          "note": 'Add new note'
        }
      }
      const state = reducer(repo, action);
      expect(state.lists[1]).toMatchSnapshot()
    });
  });
});