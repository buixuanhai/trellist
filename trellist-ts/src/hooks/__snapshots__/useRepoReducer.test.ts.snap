// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`useRepoReducer add card 1`] = `
Object {
  "cards": Array [
    Object {
      "id": "y",
      "stage": "confirmed",
      "text": "Vulnerability 3",
    },
    Object {
      "id": "z",
      "stage": "confirmed",
      "text": "Vulnerability 4",
    },
    Object {
      "id": "t",
      "stage": "confirmed",
      "text": "new card",
    },
  ],
  "id": "b",
  "stage": "confirmed",
  "title": "Confirmed",
}
`;

exports[`useRepoReducer delete card 1`] = `
Object {
  "cards": Array [
    Object {
      "id": "y",
      "stage": "confirmed",
      "text": "Vulnerability 3",
    },
  ],
  "id": "b",
  "stage": "confirmed",
  "title": "Confirmed",
}
`;

exports[`useRepoReducer move card in list 1`] = `
Object {
  "cards": Array [
    Object {
      "id": "x",
      "stage": "open",
      "text": "Vulnerability 2",
    },
    Object {
      "id": "w",
      "stage": "open",
      "text": "Vulnerability 1",
    },
  ],
  "id": "a",
  "stage": "open",
  "title": "Open",
}
`;

exports[`useRepoReducer test move card between lists 1`] = `
Object {
  "id": "1",
  "lists": Array [
    Object {
      "cards": Array [
        Object {
          "id": "x",
          "stage": "open",
          "text": "Vulnerability 2",
        },
      ],
      "id": "a",
      "stage": "open",
      "title": "Open",
    },
    Object {
      "cards": Array [
        Object {
          "id": "y",
          "stage": "confirmed",
          "text": "Vulnerability 3",
        },
        Object {
          "id": "z",
          "stage": "confirmed",
          "text": "Vulnerability 4",
        },
      ],
      "id": "b",
      "stage": "confirmed",
      "title": "Confirmed",
    },
    Object {
      "cards": Array [
        Object {
          "id": "m",
          "stage": "falsePositive",
          "text": "Vulnerability 5",
        },
        Object {
          "id": "n",
          "stage": "falsePositive",
          "text": "Vulnerability 6",
        },
      ],
      "id": "c",
      "stage": "falsePositive",
      "title": "False Positive",
    },
    Object {
      "cards": Array [
        Object {
          "id": "w",
          "stage": "fixed",
          "text": "Vulnerability 1",
        },
        Object {
          "id": "o",
          "stage": "fixed",
          "text": "Vulnerability 7",
        },
        Object {
          "id": "p",
          "stage": "fixed",
          "text": "Vulnerability 8",
        },
      ],
      "id": "d",
      "stage": "fixed",
      "title": "Fixed",
    },
  ],
  "name": "Repo 1",
}
`;

exports[`useRepoReducer update card update card note  1`] = `
Object {
  "cards": Array [
    Object {
      "id": "y",
      "stage": "confirmed",
      "text": "Vulnerability 3",
    },
    Object {
      "id": "z",
      "note": "Add new note",
      "stage": "confirmed",
      "text": "Vulnerability 4",
    },
  ],
  "id": "b",
  "stage": "confirmed",
  "title": "Confirmed",
}
`;

exports[`useRepoReducer update card update card title 1`] = `
Object {
  "cards": Array [
    Object {
      "id": "y",
      "stage": "confirmed",
      "text": "Vulnerability 3",
    },
    Object {
      "id": "z",
      "stage": "confirmed",
      "text": "Vulnerability 4 (UPDATED)",
    },
  ],
  "id": "b",
  "stage": "confirmed",
  "title": "Confirmed",
}
`;
